Add a library for configuration of clangd contexts

cdamus opened the pull request:
#### What it does

Adds a Javascript library published as an NPM package `@theia/clangd-contexts` for management of the following content in one or more `.clangd` files in a C/C++ VS Code workspace:

- Compilation database (reference to a `compile_commands.json` file)
- Flags to add to what are configured in the compilation database when invoking clang
- Flags to remove from what are configured in the compilation when invoking clang

The usage of this library is demonstrated by two example packages:

- A simple VS Code extension implementing a workspace-wide build configuration selector in the status bar
- A full-featured `clangd-context` CLI tool providing terminal access to all of the library’s capabilities
- A simple example C project with multiple clangd contexts in which to exercise the other examples

The VS Code extension example is deployed into the browser and Electron examples already provided by the Theia CPP Extensions project.

Resolves #142. 

Contributed on behalf of STMicroelectronics.

#### How to test

Build the monorepo, then run the clangd contexts library unit tests with

```console
$ yarn --cwd packages/clangd-contexts test
```

Follow the readme guides in the example packages for details of how to exercise them to perform ad hoc testing of the library:

- examples/clangd-workspace/README.md
- examples/clangd-contexts-ext/README.md
- examples/clangd-contexts-cli/README.md

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

cdamus commented:
Ah, I see some test failures on Windows platform owing to the difference in path separators. I'm working on that.

cdamus commented:
Commit 8134ced20be4bde3cdbb2419c4d3cb6cbfeb0b76 fixes the test failures on Window platform by making two changes:

- use `/` as the separator for hierarchical context names on all platforms, including Windows
- in the tests, convert paths to platform separators in assertions

When resolving context names to directories in the filesystem, the `path.resolve` and `path.join` APIs handle conversion of `/` to `\` as necessary.

JonasHelming commented:
@colin-grant-work, I was told you might be interested in this

benjaminhdavis commented:
This looks like an interesting contribution, I'm eager to try it. For adding and removing flags have you considered using [clangd/config.yaml](https://clangd.llvm.org/config)?

cdamus commented:
> This looks like an interesting contribution, I'm eager to try it. For adding and removing flags have you considered using [clangd/config.yaml](https://clangd.llvm.org/config)?

Thanks @benjaminhdavis for your interest! I look forward to your feedback.

What this API and example tooling does is essentially to automate/assist the management of those `.clangd` YAMLs. Just the subset that references the compilation database and the flags to add/remove.

colin-grant-work reviewed (commented):
I'll keep looking over the code today, but I had some comments while reading over the markdown.

colin-grant-work commented on packages/clangd-contexts/src/clangd-config.ts:
```diff
@@ -0,0 +1,102 @@
+import { isFile, toConfigPath } from './util/file-util';
+
+/** Description of a project configuration as encoded in a `.clangd` file. */
+export interface ClangdConfig {
+    CompileFlags: CompileFlags;
+    SourceFile: string;
+}
+
+/** Representation of the compile flags property of the `.clangd` configuration. */
+export interface CompileFlags {
+    Add?: string | string[];
+    Remove?: string | string[];
+    CompilationDatabase?: string;
+}
```
By convention, the members of these interfaces should have camelCase names rather than PascalCase.

cdamus replied:
The source YAML files use the Pascal case, yes. The only property that is not in the Yaml is the `SourceFile` of the `ClangdConfig`, which stores the path to the Yaml file, itself. So that one needn't be Pascal case.

colin-grant-work commented on packages/clangd-contexts/src/clangd-config.ts:
```diff
@@ -0,0 +1,102 @@
+    if (!isFile(configurationFile)) {
+        return undefined;
+    }
+    const config = yaml.load(fs.readFileSync(configurationFile, 'utf8'));
```
Could `yaml.load` throw and error if the file is incorrectly formatted?

colin-grant-work commented on packages/clangd-contexts/src/clangd-config.ts:
```diff
@@ -0,0 +1,102 @@
+/** Description of a project configuration as encoded in a `.clangd` file. */
+export interface ClangdConfig {
+    CompileFlags: CompileFlags;
+    SourceFile: string;
```
`SourceFile` has a bit of an odd ring in this context, since for a compiler that could mean the source code. Perhaps `configFile`?

colin-grant-work commented on packages/clangd-contexts/src/clangd-config.ts:
```diff
@@ -0,0 +1,102 @@
+}
+
+/**
+ * If a `.clangd` configuration file at the given path exists, load it into a {@link ClangdConfig},
+ * or else create a new empty {@link ClangdConfig}. A new empty file is not created; be sure not
+ * to {@link save} an empty configuration using the result.
```
Everywhere this function is called, the result is eventually saved. The `save` function logs an error (should it `return` or `throw` at that point?) if no source file is defined, but the `createClangdConfig` function requires a source file to run, so it appears that this function will always return a result that is well formed according to the expectations of the `save` function. What is the risk of saving a file with an empty configuration?

cdamus replied:
In general, the result will have some configuration update written into it, so it won't be empty. But it may effectively lose data if there was a file that just couldn't be read. I'm updating the `load` function to throw, but then we must decide whether this `loadOrCreate` would suppress that exception or propagate it. I lean towards the latter to prevent unexpected data loss due simply to innocently malformed YAML.

colin-grant-work commented on packages/clangd-contexts/src/clangd-config.ts:
```diff
@@ -0,0 +1,102 @@
+ */
+export function save(config: ClangdConfig): void {
+    const sourceFile = config.SourceFile;
+    if (!sourceFile) {
+        console.error('Could not save config. No "SourceFile" has been defined.');
+    }
```
`return;`? `throw`? Otherwise, the process proceeds under conditions that seem guaranteed to cause failure.

cdamus replied:
I've added a `ClangConfigException` class to throw here and in the `load[OrCreate]` APIs. I've added a test for the exception thrown in loading an invalid YAML. But I've not updated the examples to do thorough exception handling. Let me know if that's a concern.

cdamus commented:
Thanks so much @colin-grant-work for this thorough review. It is greatly appreciated!

I shall have to take some time to consider and integrate this feed-back. Stay tuned for an update anon.

cdamus commented on packages/clangd-contexts/src/clangd-contexts-config.ts:
```diff
@@ -0,0 +1,198 @@
+        this.projects.forEach(proj => {
+            const projPath = path.resolve(this.path, proj.path);
+
+            switch (proj.contextDirs) {
+                case 'flat':
+                    this.collectFlatClangdContexts(projPath, result);
+                    break;
+                case 'nested':
+                    this.collectNestedClangdContexts(projPath, result);
+                    break;
+            }
```
When more cases are added, it could be made into a switch.

cdamus commented on packages/clangd-contexts/src/util/file-util.ts:
```diff
@@ -0,0 +1,90 @@
+ * @param path a path to test
+ * @returns `true` if the `path` exists, `false` otherwise
+ */
+export function doesExist(path: string): boolean {
+    try {
+        return fs.existsSync(path);
+    } catch (error) {
+        return false;
+    }
+}
```
Nope, none.

cdamus commented:
Thanks again, @colin-grant-work for the review. Especially the copy editing of documentation! Nice to see such attention to detail. That prompted me to do another review of my own and make some further edits.

I've force-pushed commit 658d83a6258bccdb743cb954bd46f22af31f7b30 that should address your comments, including partially the matter of handling of the initial '-' of compile flags. I've updated the library to require flags verbatim and moved special handling of the initial flag-without-hyphen to the CLI example. As this is just an example of how to use the library, I'm not sure how much effort we want to invest in its look-and-feel. Certainly that's important to the library. If you feel strongly about this in the CLI, I'll be happy to take another go at it.

colin-grant-work commented on examples/clangd-contexts-ext/src/build-configuration.ts:
```diff
@@ -0,0 +1,135 @@
+ * @returns the unique root folders
+ */
+function collectUniqueRootDirs(workspaces: readonly vscode.WorkspaceFolder[]): string[] {
+    const result = workspaces.map(w => w.uri.path);
+
+    for (let i = 0; i < result.length; i++) {
+        const next = normalizeDir(result[i]);
+        for (let j = 0; j < result.length; j++) {
+            if (i !== j) {
+                const other = normalizeDir(result[j]);
+                if (next.startsWith(other)) {
+                    result.splice(j, 1);
+                }
+            }
+        }
+    }
+    return result;
```
```suggestion
    const normalized = workspaces.map(w => normalizeDir(w.uri.path));
    return normalized.filter(w => !normalized.some(otherW => otherW !== w && otherW.startsWith(w)));
```

colin-grant-work commented on examples/clangd-contexts-ext/src/ignore-gcc-flags-command.ts:
```diff
@@ -0,0 +1,43 @@
+export async function ignoreGCCFlagsCommandHandler(args: unknown[]): Promise<void> {
+    let uri = getUri(args)?.toString().replace('file://', '');
+    if (!uri && vscode.workspace.workspaceFolders) {
+        uri = vscode.workspace.workspaceFolders[0].uri.path.replace('file://', '');
```
```suggestion
        uri = vscode.workspace.workspaceFolders[0].uri.fsPath;
```

colin-grant-work commented on examples/clangd-contexts-ext/src/util.ts:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    const files = fs.readdirSync(startPath);
+    for (let i = 0; i < files.length; i++) {
+        const filename = path.join(startPath, files[i]);
+        const stat = fs.lstatSync(filename);
+        if (stat.isDirectory()) {
+            collectMatchingFiles(filename, predicate, results);
+        } else if (stat.isFile() && predicate(filename)) {
+            results.push(filename);
+        }
+    }
```
Any reason not to use `.forEach` here?

colin-grant-work commented on examples/clangd-workspace/README.md:
```diff
@@ -0,0 +1,76 @@
+The compilation commands in the `compile_commands.json` databases of some build configurations can include flags supported by GCC that the clang toolchain used by clangd does not understand.
+The clangd configuration can specify flags that should be suppressed in the invocation of the clang compiler:
+
+-   e.g. `-fstack-usage` is a GCC flag used when compiling, but yields an error for `clangd` because it is not supported by the clang compiler
```
```suggestion
-   e.g. `-fstack-usage` is a GCC flag used when compiling that yields an error for `clangd` because it is not supported by the clang compiler
```

colin-grant-work reviewed (commented):
This looks pretty good. I think the extension, and maybe the CLI source code, escaped my attention last time, so a few comments there.

cdamus commented on README.md:
```diff
@@ -18,43 +18,102 @@
+- `theia-clangd-contexts-ext`
+    - An example VS Code extension demonstrating use of the `@theia/clangd-contexts` API.
+    - See the [package readme][ccvsx] for details, including a step-by-step guide to the extension UI.
+- `@theia/cpp` **deprecated** and **removed** ([past sources][cpp]):
+    - **Deprecated** C/C++ language-features extension.\
+        (Language-features should instead be contributed by VS Code extensions).
```
I don't know why this is here or who might miss it.

This item isn't related to the feature at hand, and removing it would lose the link to past sources. Perhaps we could raise a new issue to remove this from the README if it now seems to have been long enough since the code in question was itself removed?

cdamus commented on examples/clangd-contexts-cli/src/validation.ts:
```diff
@@ -0,0 +1,196 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ *******************************************************************************/
+
```
Yes, I think these name originate from their usage in the CLI framework as processor functions to transform values parsed from the command-line. We're using them here for validation more than transformation, so they could be better named. I'll find some new names.

cdamus commented on examples/clangd-contexts-ext/README.md:
```diff
@@ -0,0 +1,69 @@
+# for the clangd example workspace opened in the browser
+$ yarn start:browser:clangd
+# for the clangd example workspace opened in the Electron deployment
+$ yarn start:electron:browser
```
Oof, good catch, thanks.

cdamus commented on examples/clangd-contexts-ext/src/extension.ts:
```diff
@@ -0,0 +1,56 @@
+}
+
+function handleFileChange(file: string): void {
+    if (file.endsWith(COMPILATION_DATABASE_FILE)) {
+        loadBuildConfiguration();
+    }
```
There's no harm in checking the precondition for correct behaviour. Especially if the watcher starts listening also for changes to other files for some reason (such as actually commented on elsewhere 😉)

cdamus commented:
Thanks again, Colin. I've made changes to address your comments on the examples and force-pushed that in commit f4b81c580db972fc76a52289ce23c49c3afd2b32.

On the matter of the activation trigger for the VS Code extension, a naïve attempt to get a `workspaceContains` trigger for compile-commands files didn't work, so I changed it instead to `onStartupFinished` so that at least it gets out of the way of VS Code's critical start-up sequence. I think `onCommand` wouldn't work because the main UI actually is a status line item? As this is just an example, not intended for production use, perhaps that will be sufficient. Otherwise, perhaps we could raise a follow-up issue to try to better narrow the activation conditions.

With the additional watcher on `.clangd` files now using the CLI to change context selection in a terminal within VS Code causes the status line to update, which is a nice touch. Thanks for prompting that improvement.

Otherwise I have more or less applied your suggestions except as indicated in replies in-line in the PR.

colin-grant-work commented:
@cdamus, I'll give this another quick round of functional testing, but it looks good to me. 👍

colin-grant-work reviewed (approved):
Functionally, things seem to be working well, and the code looks good to me.

vince-fugnitto reviewed (commented):
@cdamus @JonasHelming wouldn't the changes proposed in the pull-request be better suited under [eclipse-cdt-cloud](https://github.com/eclipse-cdt-cloud) directly (possibly in their own repo(s)) rather than trying to fix it in this repo. The repo as is is meant to contain the sources of `@theia` extensions related to C/C++ development. AFAIK the library added is not necessarily related to `theia` and would be clearer on it's own. Moreover, each of the components now introduced by the pull-request have their own ways to be built and packaged, and do not share a common release strategy which becomes confusing.

JonasHelming commented:
> @cdamus @JonasHelming wouldn't the changes proposed in the pull-request be better suited under [eclipse-cdt-cloud](https://github.com/eclipse-cdt-cloud) directly (possibly in their own repo(s)) rather than trying to fix it in this repo. The repo as is is meant to contain the sources of `@theia` extensions related to C/C++ development. AFAIK the library added is not necessarily related to `theia` and would be clearer on it's own. Moreover, each of the components now introduced by the pull-request have their own ways to be built and packaged, and do not share a common release strategy which becomes confusing.

Yes, this is correct. When we started the PR, CDT.cloud did not exist. I would merge this now so we have the contribution legally finished and then move it to CDT.cloud. Fine for you?

JonasHelming commented:
@vince-fugnitto ?

vince-fugnitto commented:
> Yes, this is correct. When we started the PR, CDT.cloud did not exist. I would merge this now so we have the contribution legally finished and then move it to CDT.cloud. Fine for you?

@JonasHelming unfortunately I'm still not sure I fully understand why the changes need to be fitted into this repo instead of directly under CDT.cloud now that it exists, perhaps I am not aware of the legal reasons. I still feel that trying to fit everything under a single repo (when they are somewhat unrelated or each component have different release strategies) is problematic, and doing this separation later will become harder. I'll leave the decision up to you however, you can merge if you feel it's necessary.

cdamus commented:
This contribution has been accepted in the CDT.Cloud project, so this PR is now obsolete. Thanks for the review! That was all to the benefit of CDT.Cloud.
